[IV-23-24] Objetivo 2

marinajcs opened the pull request:
Enlace a [wiki](https://github.com/marinajcs/maybe-better-maybe-worse/wiki/Esquema-ideas-%5BHU01%5D) de apoyo.

## Sobre la estructura del repositorio
* [x] ¿He seguido las mejores prácticas en el nombre de las clases y ficheros y
   disposición de los mismos?
* [x] ¿He mirado tanto reglas generales (para cualquier repo) como reglas
específicas para el lenguaje con el que se está trabajando?

## Sobre el planteamiento
* [x] ¿Se han planteado una serie de issues, asignados a las historias de
usuario pertinentes, que clarifiquen qué es lo que necesitan los usuarios e
identifiquen las diferentes partes del problema?
* [x] ¿Los issues representan un problema, y no una tarea?

## Sobre el análisis del problema
* [x] ¿Se ha documentado qué análisis se ha hecho sobre el dominio para decir lo
   que se ha creado?
* [x] ¿Se ha documentado por qué se ha elegido que lo creado sea un objeto valor,
   una entidad o un agregado?
* [x] ¿He propuesto un producto mínimamente viable, que en muchos casos será un solo
   objeto valor que no dependa de ningún otro (y que sea la base de muchos
   otros)?

## Sobre el código
* [x] ¿Todos los mensajes de commit explican el cambio, y no se
      limitan a repetir el nombre del fichero que se ha cambiado?
* [x] ¿Los mensajes de commit siguen el formato estándar y buenas prácticas?
* [x] ¿Se ha hecho una revisión real del código para comprobar que todos
   los atributos y funciones creadas están respaldadas por una HU?
* [x] ¿Todos los issues creados están relacionados con a una HU?
* [x] ¿Ha asignado el propietario/la propietaria del repositorio todos los
issues planteados al *milestone* correspondiente?
* [x] ¿Todos los cambios en el código están asignados a un issue al que se
      referencia en un commit?
* [x] ¿Se ha asignado al mismo milestone el PR que se ha hecho?
* [x] ¿Es el milestone sobre el que estoy trabajando un producto
   mínimamente viable? ¿O tengo que solicitar al product manager que precise de
   qué producto se trata?
* [x] ¿Se ha comprobado que el código entregado sea sintácticamente
correcto?

marinajcs committed:
Creado fichero yaml con lenguaje acordado closes #10

marinajcs committed:
Corregido lenguaje de iv.yaml

marinajcs committed:
Modificado .gitignore para lenguaje Go

marinajcs committed:
Añadida entidad en iv.yaml

marinajcs committed:
Creada entidad plan de rodaje closes #12

marinajcs committed:
Creada entidad plano closes #13

marinajcs committed:
Creada entidad actor closes #14

marinajcs committed:
Añadidos los constructores de Actor, Plano y PlanRodaje

diagmatrix commented on src/actor.go:
```diff
@@ -0,0 +1,38 @@
+import "time"
+
+// Actor representa un actor con información básica y su disponibilidad horaria.
+type Actor struct {
```
Creo que la edad del actor no es necesaria, por lo demás me parece perfecta la definición de `Actor`

marinajcs replied:
Corregido en https://github.com/diagmatrix/maybe-better-maybe-worse/pull/11/commits/addf0dd910d71b6b0dc0e7a95e4241edbc501267

diagmatrix commented on src/planRodaje.go:
```diff
@@ -0,0 +1,40 @@
+package src
+
+// PlanRodaje representa un plan de rodaje que contiene una serie de planos
+type PlanRodaje struct {
```
Aquí, cambiaría los campos de dirección y producción por `[]string`, ya que puede ser un producto co-dirigido o con varios productores.

marinajcs replied:
Corregido en https://github.com/diagmatrix/maybe-better-maybe-worse/pull/11/commits/33fe0b71a31cf7a82d9920a6275b2c117680ace2

diagmatrix reviewed (changes requested):
Salvo por algunos parámetros que he comentado, me parece muy buena modelización del problema

marinajcs committed:
Modificadas variables de clase en planRodaje.go

Las variables dirección y producción se han convertido en un slice de string, ya que puede ser un plan de rodaje dirigido/producido por más de una persona.

marinajcs committed:
Eliminada variable de edad en actor.go

diagmatrix commented:
* [x] ¿Los issues que ha planteado la persona son efectivamente problemas y
evidencian comprensión de las historias de usuario? Si no es así, es mejor que
se le transmita antes de que empiece a crear cualquier código erróneo.
* [x] ¿Has comprobado que efectivamente el compañero/a ha seguido *su* lista de
      comprobación?
* [x] ¿El código creado es una modelización válida de las historias de usuario
      creadas por ti, y te sirve como base para programar la lógica de negocio?
* [x] ¿El código creado sigue las mejores prácticas para el lenguaje de
      programación que hayáis consensuado, tanto en identificadores, como
      nombres de ficheros, como disposición de los diferentes ficheros en el
      repositorio? Por ejemplo, ¿todos los commits se refieren a un issue y
      aportan código para resolver el problema planteado en el mismo?
* [x] ¿He comprobado que el código compile, o que quien lo ha programado ha lo
ha hecho y documentado en el repositorio?
* [x] ¿He añadido el PR y todos los issues que se resuelvan en él al milestone
correspondiente, que será el primero?
* [x] ¿Cubre totalmente el milestone 0, o hará falta modelización adicional en
el objetivo 4 para poder programar la lógica de negocio?
* [x] En resumen, ¿considero que con lo hecho puedo ponerme a trabajar
directamente en el siguiente milestone, que incluye la lógica de negocio?

diagmatrix commented:
Listo para revisión @JJ

JJ commented:
> ## Sobre el planteamiento
> 
>     * [x]  ¿Se han planteado una serie de issues, asignados a las historias de
>       usuario pertinentes, que clarifiquen qué es lo que necesitan los usuarios e
>       identifiquen las diferentes partes del problema?

En issues como el #12 no has enlazado la historia de usuario, sino una descripción del problema. El enlazar una historia de usuario no es arbitrario; GitHub mostrará las HUs que están enlazadas con tal historia y se podrá navegar de historia a código y al revés.

Por supuesto, enlazar la HU muestra que se está trabajando en tal historia y por tanto aportando valor al cliente. En tu  caso has enlazado descripciones en el repo, así que asumo que esa parte sí la tienes cubierta, pero habría que cubrirla toda. Además, hay muchas historias de usuario. Si no dices claramente cuál es la que está vigente, es imposible saber si el código es correcto o no.

JJ commented:
Se deben respetar también las buenas prácticas en los mensajes de commits, imprescindibles para la documentación del código. Como dice en [el fichero de errores frecuentes](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-2.md#errores-frecuentes-objetivo-2), es un error...

> Poner en el mensaje de commit "Solución xx". Decir que es la solución no contribuye a entender por qué se trata de la solución. Tampoco "Añadido x". El mensaje debe añadir información a lo que ya es evidente, porque en la información del mismo se ve claramente qué se ha añadido o modificado. Debe explicar por qué se ha tomado esa decisión específica de crear ese objeto para resolver el problema y no otro.

Esto ya no tiene solución, pero os rogaría que lo tuviérais en cuenta en el futuro. Las buenas prácticas aseguran en todo caso que el código sea correcto y añada valor al cliente; no seguirlas es un buen predictor de que no va a suceder así.

JJ commented on iv.yaml:
```diff
@@ -0,0 +1,2 @@
+lenguaje: Go
+entidad: src/planRodaje.go
```
Como hemos dicho repetidamente en clase, no se pueden usar mayúsculas en un repo por la carencia de cierrtos sistemas operativos a distinguir mayúsculas y minúsculas.

marinajcs replied:
Corregido en https://github.com/diagmatrix/maybe-better-maybe-worse/pull/11/commits/891bfd6a92f72f1a8584642f7bfe5f0bd66d85cf

JJ commented on src/actor.go:
```diff
@@ -0,0 +1,37 @@
+package src
```
¿Has consultado las mejores prácticas en el directorio que se debe usar el Go? ¿Y en el nombre del package? En el grupo de Telegram de clase se contestó una pregunta al respecto; te agradecería que la consultaras.

marinajcs replied:
Corregido en https://github.com/diagmatrix/maybe-better-maybe-worse/pull/11/commits/47f1ad4657b750125f3f92c40021d49dce1fa2a9

JJ commented on src/actor.go:
```diff
@@ -0,0 +1,37 @@
+
+// Actor representa un actor con información básica y su disponibilidad horaria.
+type Actor struct {
+	dni            string      // Número de identificación del actor
```
Precisamente lo que te comentaba en el comentario sobre seguir buenas prácticas aquí se hace evidente. Te ha comentado @diagmatrix en [este comentario](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/11/files#r1374805240) que no hace falta la edad del actor. Lo has cerrado en el commit https://github.com/diagmatrix/maybe-better-maybe-worse/commit/addf0dd910d71b6b0dc0e7a95e4241edbc501267, *que no referencia ningún issue*. Pero es que el código que lo creó, 47e0886, tampoco sigue las mejores prácticas porque no dice por qué el DNI hace falta para cuadrar la disponibilidad de los actores. En el [fichero de errores frecuentes](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-2.md#sobre-los-objetos-valor) comenta que:

> Con el objeto usuario es con el que más errores se cometen. Para empezar en muchas ocasiones y en un PMV ni siquiera hace falta una clase, es simplemente un valor único. Pero para seguir, sólo hay que incluir lo que vaya a usarse en la lógica de negocio. Ni password, ni DNI, ni nombre separado de apellidos... sólo lo que se pida en la lógica de negocio.

marinajcs replied:
Corregido en https://github.com/diagmatrix/maybe-better-maybe-worse/pull/11/commits/15b13d4dd268b733987cb852721eb3e323b08e4f

JJ reviewed (changes requested):
Es un PR bastante completo, y se agradece sobre todo la documentación de cara al usuario. Sin embargo, hay una serie de problemas de buenas prácticas, de formato pero también de planteamiento. Adicionalmente a lo comentado, tened en cuenta [este otro error frecuente](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-2.md#sobre-el-c%C3%B3digo-definido):

> Si el constructor no añade nada a la clase o struct definido, es mejor no hacerlo; por ejemplo, si simplemente inicializa las variables con respecto a lo que se le pasa como argumento.

JJ commented:
¿Alguna aclaración necesaria aquí? Sería conveniente que avanzárais con esto para terminarlo lo antes posible.

marinajcs committed:
Eliminada mayúscula del nombre del fichero planrodaje.go

Algunos sistemas operativos no distinguen mayúsculas y minúsculas, por lo que se debe evitar usarlas

marinajcs committed:
Cambiada nomenclatura del directorio y los paquetes de src a internal

Se sigue la práctica ampliamente aceptada de crear el directorio internal en Go, que se utiliza para crear paquetes que solo deben ser accesibles dentro del proyecto

marinajcs committed:
Eliminados constructores que no añaden a los structs de actor, plano y planrodaje

Simplemente inicializaban los valores que se le pasaban por parámetros y no incluían comportamientos adicionales

marinajcs committed:
Eliminadas variables de actor no imprescindibles, fixes #14

Se han suprimido las variables de nombre y DNI no necesarias para la implementación de la lógica de negocio. Adicionalmente, se han eliminado comentarios triviales.

marinajcs committed:
Eliminados comentarios triviales en documentación de plano, fixes #13

marinajcs committed:
Suprimidas variables no imprescindibles en plan de rodaje, fixes #12

Se han eliminado las variables: código, película, dirección y producción, ya que no son necesarias para la lógica de negocio ni se piden en las historias de usuario

marinajcs committed:
Corregido formato de struct Actor, fixes #14

El comentario no se correspondía a lo incluido explícitamente en el código y había un pequeño error en el espaciado entre variable y tipo

marinajcs committed:
Eliminado comentario de struct Plano, fixes #13

Los nombres del struct y sus variables son suficientes para documentar la propia entidad

marinajcs committed:
Añadido struct que asigna un plano a una fecha, fixes #12

Para poder implementar bien la lógica de negocio, un plan de rodaje debe incluir un conjunto de planos, cada uno asignado a una fecha concreta. Al no existir Pair en Go, se ha declarado un struct para dichas asignaciones.

marinajcs committed:
Arreglado pequeño fallo de espaciado en planrodaje.go, fixes #12

marinajcs commented:
Listo para revisión @JJ

JJ reviewed (approved):
👍

JJ commented:
Convendría que comenzaras con el objetivo 3 lo antes posible, para entregarlo la próxima semana preferiblemente
